docs: add release note for 0.4.0 (#531)
web-infra-dev/rspack-website: docs/en/blog/announcing-0.4.mdx

# Rspack 0.4 Release Announcement

## Announcing Rsbuild v0.1

> ![](https://p3-juejin.byteimg.com/tos-cn-i-k3u1fbpfcp/a0fe4dcdc2264c6f81a72ddeabd51268~tplv-k3u1fbpfcp-jj-mark:0:0:0:0:q75.image#?w=1800&h=579&s=145010&e=png&b=fffdf1)

Author: should use local image

We are pleased to announce **the release of** **[Rsbuild](https://github.com/web-infra-dev/rsbuild)** **v0.1!** This milestone indicates that the Rsbuild API is now more stable and reliable to use.

> Rsbuild is an Rspack-based build tool, designed to be **an enhanced Rspack** **CLI** that is both more user friendly and powerful. As a result of the Rspack team's exploration and implementation of best practices for web building, Rsbuild is the ideal solution for those looking to migrate from Webpack to Rspack. It significantly reduces configuration by 90% while offering a 10x build speed.

Reviewer: ```suggestion
Rsbuild is an Rspack-based build tool, designed to be **an enhanced Rspack** **CLI** that is both more user friendly and powerful. As a result of the Rspack team's exploration and implementation of best practices for web building, Rsbuild is the ideal solution for those looking to migrate from Webpack to Rspack. It significantly reduces configuration by 90% while offering a 5~10x build speed.
```

## Major Changes

### Drop NodeJS 14 Support

Rspack no longer supports Node.js 14, Node.js 16+ is now required.

### Make @rspack/core as peerDep of @rspack/cli

`@rspack/core` is now a peerDependency of `@rspack/cli` rather than a direct dependency. This means that you need to manually install `@rspack/core` with `@rspack/cli` now. aligning Rspack more closely with webpack. In the long term, the positioning of `@rspack/cli` will no longer be an out-of-the-box solution. We will align `@rspack/cli` with webpack-cli and may even directly support the use of `@rspack/core` in `webpack-cli`. We recommend [Rsbuild](https://rsbuild.dev/) as an out-of-the-box solution.

### Deprecating default transformation

`experiments.rspackFuture.disableTransformByDefault` is enabled by default in v0.4.0. For people that still need the legacy behavior, you may manually set this option to `false`.

This feature primarily addresses three categories of problems: [builtins](https://www.rspack.dev/config/builtins) code transformation features, [target](https://www.rspack.dev/config/target), and custom [Rule.type](https://www.rspack.dev/config/module#ruletype).

> 1.  Removal of support for some [[builtins](https://www.rspack.dev/config/builtins)](https://www.rspack.dev/config/builtins) features:
> 
> - [[builtins.relay](https://www.rspack.dev/config/builtins#builtinsrelay)](https://www.rspack.dev/config/builtins#builtinsrelay): moved to `rspackExperiments.relay`
> - [[builtins.react](https://www.rspack.dev/config/builtins#builtinsreact)](https://www.rspack.dev/config/builtins#builtinsreact): moved to `jsc.transform.react`
> - [[builtins.emotion](https://www.rspack.dev/config/builtins#builtinsemotion)](https://www.rspack.dev/config/builtins#builtinsemotion): moved to `rspackExperiments.emotion`
> - [[builtins.pluginImport](https://www.rspack.dev/config/builtins#builtinspluginimport)](https://www.rspack.dev/config/builtins#builtinspluginimport): moved to `rspackExperiments.import`
> - [[builtins.decorator](https://www.rspack.dev/config/builtins#builtinsdecorator)](https://www.rspack.dev/config/builtins#builtinsdecorator): moved to `jsc.parser.decorator`
> - [[builtins.presetEnv](https://www.rspack.dev/config/builtins#builtinspresetenv)](https://www.rspack.dev/config/builtins#builtinspresetenv): moved to `jsc.env`

Reviewer 2: ```suggestion
1.  Removal of support for some [builtins](https://www.rspack.dev/config/builtins) features:

- [builtins.relay](https://www.rspack.dev/config/builtins#builtinsrelay): moved to `rspackExperiments.relay`
- [builtins.react](https://www.rspack.dev/config/builtins#builtinsreact): moved to `jsc.transform.react`
- [builtins.emotion](https://www.rspack.dev/config/builtins#builtinsemotion): moved to `rspackExperiments.emotion`
- [builtins.pluginImport](https://www.rspack.dev/config/builtins#builtinspluginimport): moved to `rspackExperiments.import`
- [builtins.decorator](https://www.rspack.dev/config/builtins#builtinsdecorator): moved to `jsc.parser.decorator`
- [[builtins.presetEnv](https://www.rspack.dev/config/builtins#builtinspresetenv)](https://www.rspack.dev/config/builtins#builtinspresetenv): moved to `jsc.env`
```

```js

module.exports = {
  module: {
    rules: [
      {
        test: /.jsx$/,
        loader: 'builtin:swc-loader',
        options: {
          jsc: {
            parser: {
              syntax: "ecmascript",
              jsx: true
            }
          },
          transform:{
            react: {
              runtime: "automatic"
            }
          },
          rspackExperiments: {
            emotion: true // The same as `builtins`
          }
        }
        type: 'javascript/auto',
      },
    ],
  },
  experiments: {
    rspackFuture: {
      disableTransformByDefault: true
    }
  }
};
```

> 2.  [[target](https://www.rspack.dev/config/target)](https://www.rspack.dev/config/target) will not downgrade user-side code

Reviewer 2: ```suggestion
2. [target](https://www.rspack.dev/config/target) will not downgrade user-side code
```

```diff
module.exports = {
  target: ["web", "es5"],
  module: {
    rules: [
      {
        test: /.js$/,
        exclude: /node_modules/,
        loader: 'builtin:swc-loader',
        options: {
          jsc: {
            parser: {
              syntax: "ecmascript"
            },
+           target: "es5" // Notice: `jsc.target` and `env` cannot be set at the same time.
          },

> +        env: { //  Notice: `jsc.target` and `env` cannot be set at the same time.
> +         targets: "chrome >= 48"
> +        }

Reviewer 2: ```suggestion
+         env: { //  Notice: `jsc.target` and `env` cannot be set at the same time.
+           targets: "chrome >= 48"
+         }
```

        }
        type: 'javascript/auto',
      },
    ],
  }
};
```

> 3.  Removed non-webpack compatible [[Rule.type](https://www.rspack.dev/config/module#ruletype)](https://www.rspack.dev/config/module#ruletype)

Reviewer 2: ```suggestion
3.  Removed non-webpack compatible [Rule.type](https://www.rspack.dev/config/module#ruletype)
```

These types have been removed:

- `"typescript"`
- `"jsx"`
- `"tsx"`

For JS-related types, only the following will be retained:

- `"javascript/auto"`
- `"javascript/esm"`
- `"javascript/dynamic"`

Refer to [this](https://www.rspack.dev/config/experiments.html#experimentsrspackfuturedisabletransformbydefault) for the complete migration guide.

Check out our previous discussion [here](https://github.com/web-infra-dev/rspack/discussions/4070).

### Deprecating builtin.react.refresh

> With `experiments.rspackFuture.disableTransformByDefault` is enabled by default in v0.4.0, `builtin.react.refresh` has also been deprecated. Now we recommend using `@rspack/plugin-react-refresh` to enable react fast refresh.

Reviewer 2: ```suggestion
With `experiments.rspackFuture.disableTransformByDefault` is enabled by default in v0.4.0, `builtin.react.refresh` has also been deprecated. Now we recommend using `@rspack/plugin-react-refresh` to enable react fast refresh. And to use it, you need install [@rspack/plugin-react-refresh](https://www.npmjs.com/package/@rspack/plugin-react-refresh) and [react-refresh](https://www.npmjs.com/package/react-refresh)
```

```diff
+ const ReactRefreshPlugin = require('@rspack/plugin-react-refresh');
const isDev = process.env.NODE_ENV === 'development';

module.exports = {
  // ...
  mode: isDev ? 'development' : 'production',
  module: {
    rules: [
      {
        test: /.jsx$/,
        use: {
          loader: 'builtin:swc-loader',
          options: {
            jsc: {
              parser: {
                syntax: 'ecmascript',
                jsx: true,
              },
              transform: {
                react: {
+                  development: isDev,
+                  refresh: isDev,
                },
              },
            },
          },
        },
      },
    ],
  },
-  builtins: {
-    react: {
-      refresh: true,
-    }
-  },
  plugins: [
+    isDev && new ReactRefreshPlugin()
  ].filter(Boolean),
};
```

Checkout [here](https://www.rspack.dev/zh/guide/react.html#fast-refresh) for more details.

### Deprecating builtin:sass-loader

`builtin:sass-loader` has now been deprecated. If you are using it, migrate to `sass-loader`. Rspack will remove `builtin:sass-loader` in v0.5.0.

> ### Deprecating experiments.incrementalRebuild

Reviewer 2: ```suggestion
### Deprecating experiments.incrementalRebuild options
```

> `experiments.incrementalRebuild` has now been deprecated. Rspack will remove it in v0.5.0.

Reviewer 2: ```suggestion
`experiments.incrementalRebuild` has now internally enabled, so you don't need the options anymore, the options has been deprecated. Rspack will remove the options in v0.5.0.
```

## Rspack 0.4 Major Changes

### Drop NodeJS 14 Support

Rspack no longer supports Node.js 14, Node.js 16+ is now required.

### Make @rspack/core as peer dependency of @rspack/cli

`@rspack/core` is now a peer dependency of `@rspack/cli` rather than a direct dependency. This means that you need to manually install `@rspack/core` with `@rspack/cli` now. aligning Rspack more closely with webpack. In the long term, the positioning of `@rspack/cli` will no longer be an out-of-the-box solution. We will align `@rspack/cli` with webpack-cli and may even directly support the use of `@rspack/core` in `webpack-cli`. We recommend [Rsbuild](https://rsbuild.dev/) as an out-of-the-box solution.

### Deprecating default transformation

`experiments.rspackFuture.disableTransformByDefault` is enabled by default in v0.4.0. For people that still need the legacy behavior, you may manually set this option to `false`.

This feature primarily addresses three categories of problems: [builtins](https://www.rspack.dev/config/builtins) code transformation features, [target](https://www.rspack.dev/config/target), and custom [Rule.type](https://www.rspack.dev/config/module#ruletype).

> 1. Removal of support for some [[builtins](https://www.rspack.dev/config/builtins) features:

Reviewer 3: ```suggestion
1. Removal of support for some [builtins](https://www.rspack.dev/config/builtins) features:
```
